tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc

```diff
@@ -155,6 +155,12 @@ Status TrtShapeOptimizationProfile::CollectShapeValues(OpKernelContext* ctx) {
   // First copy all the shape value candidates into actual_shape_values_ vector.
   for (int i = 0; i < ctx->num_inputs(); i++) {
     if (is_shape_tensor_[i]) {
+      if (ctx->input_dtype(i) != DT_INT32) {
+        // In case the is_shape_tensor mask was initialized with the input
+        // shapes only (without knowledge of dtype) then we apply correction.
```

Comment:
Can you point me to such initialization? I meant, we initialize the value to "true" unsafely?
What I can tell is we either initialize it to false or IsTrtShapeTensorCompatible.

Reply:
`InTrtShapeTensorCompatible` has an overload which takes only shape, but no dtype information. We have one version of [SetShapeTensorMask](https://github.com/tensorflow/tensorflow/blob/63307afb2814e7df213140ecfb30bbcb5a861c9c/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc#L405) that uses that. (See also discussion below for the test added in this PR). Until the TRT network is defined we have only a guess of the `is_shape_tensor` values. Yes, the existing implementation was setting it to `true` unsafely.

- The sanity check here adjust the `is_shape_tensor` with the dtype information. This will make the guess a "safe".
- This PR also turns off this unsafe initialization `SetShapeTensorMask` if `is_shape_tensor` is already defined.